warp-tls/Network/Wai/Handler/WarpTLS.hs

```diff
@@ -71,17 +73,52 @@ import System.IO.Error (isEOFError, ioeGetErrorType)
 
 ----------------------------------------------------------------
 
--- | Settings for WarpTLS.
-data TLSSettings = TLSSettings {
+-- | Determines where to load the certificate, chain 
+-- certificates, and key from.
+data CertSettings 
+  = CertFromFile {
```

Comment:
I'd been thinking of something slightly different, more like:

```haskell
data BytesSource = FromFile !FilePath | FromMemory !ByteString | FromRef !(IORef ByteString)

data TLSSettings = TLSSettings
  { ...
  , certSource :: !BytesSource
  , chainCertsSource :: ![BytesSource]
  , keySource :: !BytesSource
  }
```

Not convinced my way is better, but it does allow a bit more flexibility in mixing and matching data sources.

Reply:
That is true. I could change this, but we would no longer be able to use `credentialLoadX509Chain`. We would then have to use `readSignedObject` directly for each file, but that doesn't return a `ByteString` and neither does `readSignedObjectFromMemory`, so we would need something like:

```haskell
certFromBytesSource :: (ASN1Object a, Eq a, Show a) => BytesSource -> IO [SignedExact a]
certFromBytesSource (FromFile fp) = readSignedObject fp
certFromBytesSource (FromMemory bs) = readSignedObjectFromMemory fp
certFromBytesSource (FromRef ref) = readIORef ref >>= readSignedObjectFromMemory

-- something similar for keys with readKeyFile and readKeyFileFromMemory

loadCredentials TLSSettings{..} = do 
  cert <- certFromBytesSource certSource
  chain <- mapM certFromBytesSource chainCertsSource
  key <- keyFromBytesSource keySource
  case keys of
    []    -> error "no keys found"
    (key:_) -> return (CertificateChain . concat $ cert : chain, key)
```

Reply:
In that case, I don't have a strong opinion either way. If you think the current approach is better, let's stick with it. I'd like to get sign-off from Kazu, but otherwise no more comments from me. Thanks!